### PR TITLE
[SC-184] Use XChainHelpers instead of custom forwarders

### PR DESCRIPTION
**Pull request by barrutko**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,16 +4,16 @@ This repository contains infrastructure necessary to process cross-chain executi
...
 ![Architecture Diagram](/diagram.png)
 ## 🤝 Contribution Guidelines
 In order to add governance relay infrastructure for a new domain, perform the following steps:
-1. Go to [XChain Helpers](https://github.com/marsfoundation/xchain-helpers) repository and add a proper domain helper abstracting away the process of passing messages between host domain and your bridged domain.
-2. Add proper `CrosschainForwarder` to the `/src/forwarders` directory and proper `BridgeExecutor` to the `/src/executors` directory. Follow currently used naming convention.
+1. Go to [XChain Helpers](https://github.com/marsfoundation/xchain-helpers) repository and add:
+    1. A domain helper abstracting away the process of passing messages between host domain and your bridged domain.
+    2. A helper function to the XChainForwarders library abstracting away host domain interactions with the bridge.
+2. Add `BridgeExecutor` to the `/src/executors` directory. Follow currently used naming convention.
 3. If there are any dependencies required, copy them into `/src/dependencies/<YOUR_DOMAIN_NAME>` directory.
```
Comment (hexonaut) [inline README.md]: I would just drop this directory and inline all remaining interfaces/libraries. `AddressAliasHelper` function that is used can just be added to the Executor as an internal function imo.

**Review by hexonaut — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (barrutko) [reply README.md]: Done ✅

**Review by hexonaut — DISMISSED** [review]
Good to merge.

File: test/mocks/GnosisReconfigurationPayload.sol [hunk 2]
```diff
@@ -3,10 +3,12 @@ pragma solidity ^0.8.10;
 
 import { GnosisBridgeExecutor } from '../../src/executors/GnosisBridgeExecutor.sol';
 
+import { IPayload } from '../interfaces/IPayload.sol';
+
 /**
  * @dev This payload reconfigures Gnosis bridge executor to a given state
  */
-contract GnosisReconfigurationPayload {
+contract GnosisReconfigurationPayload is IPayload {
```
Comment (lucas-manuel) [inline test/mocks/GnosisReconfigurationPayload.sol:11]: Add `override` keyword to `execute()` function below

File: test/mocks/PayloadWithEmit.sol [hunk 3]
```diff
@@ -1,10 +1,12 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.10;
 
+import { IPayload } from '../interfaces/IPayload.sol';
+
 /**
  * @dev This payload simply emits an event on execution
  */
-contract PayloadWithEmit {
+contract PayloadWithEmit is IPayload {
```
Comment (lucas-manuel) [inline test/mocks/PayloadWithEmit.sol:9]: Add `override` keyword to `execute` below

File: test/mocks/ReconfigurationPayload.sol [hunk 4]
```diff
@@ -3,10 +3,12 @@ pragma solidity ^0.8.10;
 
 import { IExecutorBase } from '../../src/interfaces/IExecutorBase.sol';
 
+import { IPayload } from '../interfaces/IPayload.sol';
+
 /**
  * @dev This payload reconfigures bridge executor to a given state
  */
-contract ReconfigurationPayload {
+contract ReconfigurationPayload is IPayload {
```
Comment (lucas-manuel) [inline test/mocks/ReconfigurationPayload.sol:11]: Add `override` keyword to `execute()` function below

File: src/executors/ArbitrumBridgeExecutor.sol [hunk 5]
```diff
@@ -11,9 +10,21 @@ import {L2BridgeExecutor} from './L2BridgeExecutor.sol';
...
+        }
+    }
+
   /// @inheritdoc L2BridgeExecutor
   modifier onlyEthereumGovernanceExecutor() override {
-    if (AddressAliasHelper.undoL1ToL2Alias(msg.sender) != _ethereumGovernanceExecutor)
+    if (undoL1ToL2Alias(msg.sender) != _ethereumGovernanceExecutor)
       revert UnauthorizedEthereumExecutor();
     _;
   }
```
Comment (lucas-manuel) [inline src/executors/ArbitrumBridgeExecutor.sol:30]: Rest of this file uses 2 spaces indentation, should follow that convention (at least for this PR imo).

Also confirmed that there is no diff between this and the code in the AddressAliasHelper.sol file that was removed.

File: test/ArbitrumCrosschainTest.t.sol [hunk 6]
```diff
@@ -4,14 +4,34 @@ pragma solidity ^0.8.0;
...
 
-import { CrosschainTestBase } from './CrosschainTestBase.sol';
+import { IPayload } from './interfaces/IPayload.sol';
+
+import { CrosschainPayload, CrosschainTestBase } from './CrosschainTestBase.sol';
+
+contract ArbitrumCrosschainPayload is CrosschainPayload {
+
+    constructor(IPayload _targetPayload, address _bridgeExecutor) CrosschainPayload(_targetPayload, _bridgeExecutor) {}
+
```
Comment (lucas-manuel) [inline test/ArbitrumCrosschainTest.t.sol:19]:
```suggestion
    constructor(IPayload _targetPayload, address _bridgeExecutor) 
        CrosschainPayload(_targetPayload, _bridgeExecutor) {}
```

File: test/ArbitrumCrosschainTest.t.sol [hunk 7]
```diff
@@ -4,14 +4,34 @@ pragma solidity ^0.8.0;
...
+        );
+    }
+
+}
 
 contract ArbitrumCrosschainTest is CrosschainTestBase  {
 
+    function deployCrosschainPayload(IPayload targetPayload, address bridgeExecutor) public override returns(IPayload) {
+        return IPayload(new ArbitrumCrosschainPayload(targetPayload, bridgeExecutor));
+    }
```
Comment (lucas-manuel) [inline test/ArbitrumCrosschainTest.t.sol:36]:
```suggestion
    function deployCrosschainPayload(IPayload targetPayload, address bridgeExecutor)
        public override returns(IPayload)
    {
        return IPayload(new ArbitrumCrosschainPayload(targetPayload, bridgeExecutor));
    }
```

File: test/ArbitrumCrosschainTest.t.sol [hunk 8]
```diff
@@ -27,31 +47,9 @@ contract ArbitrumCrosschainTest is CrosschainTestBase  {
...
 
         hostDomain.selectFork();
-        forwarder = address(new CrosschainForwarderArbitrum(bridgeExecutor));
         vm.deal(
             L1_EXECUTOR,
             0.01 ether
         );
     }
-
-    function testFuzz_arbitrumGasCalculations(uint256 payloadDataLength) public {
```
Comment (lucas-manuel) [inline test/ArbitrumCrosschainTest.t.sol:37]: Looks like theres no test coverage of this in xchain helpers, am I missing something?

**Review by lucas-manuel — COMMENTED** [review]
Midway through review

File: test/CrosschainTestBase.sol [hunk 9]
```diff
@@ -24,6 +24,44 @@ struct L2BridgeExecutorArguments {
...
+
+    constructor(IPayload _targetPayload, address _bridgeExecutor) {
+        targetPayload  = _targetPayload;
+        bridgeExecutor = _bridgeExecutor;
+    }
+
+    function execute() external virtual;
+
+    function encodeCrosschainExecutionMessage() internal view returns (bytes memory) {
+
```
Comment (lucas-manuel) [inline test/CrosschainTestBase.sol]: Rm newline

File: test/CrosschainTestBase.sol [hunk 10]
```diff
@@ -24,6 +24,44 @@ struct L2BridgeExecutorArguments {
...
+        address[] memory targets = new address[](1);
+        targets[0] = address(targetPayload);
+        uint256[] memory values = new uint256[](1);
+        values[0] = 0;
+        string[] memory signatures = new string[](1);
+        signatures[0] = 'execute()';
+        bytes[] memory calldatas = new bytes[](1);
+        calldatas[0] = '';
+        bool[] memory withDelegatecalls = new bool[](1);
+        withDelegatecalls[0] = true;
```
Comment (lucas-manuel) [inline test/CrosschainTestBase.sol:49]:
```suggestion
        address[] memory targets    = new address[](1);
        uint256[] memory values     = new uint256[](1);
        string[]  memory signatures = new string[](1);
        bytes[]   memory calldatas  = new bytes[](1);

        bool[] memory withDelegatecalls = new bool[](1);

        targets[0]    = address(targetPayload);
        values[0]     = 0;
        signatures[0] = 'execute()';
        calldatas[0]  = '';

        withDelegatecalls[0] = true;
```

File: test/CrosschainTestBase.sol [hunk 11]
```diff
@@ -24,6 +24,44 @@ struct L2BridgeExecutorArguments {
...
+            targets,
+            values,
+            signatures,
+            calldatas,
+            withDelegatecalls
+        );
+    }
+
+}
+
```
Comment (lucas-manuel) [inline test/CrosschainTestBase.sol:62]: Extra newline here

File: test/GnosisCrosschainTest.t.sol [hunk 12]
```diff
@@ -4,24 +4,43 @@ pragma solidity ^0.8.0;
...
 
 import { GnosisReconfigurationPayload } from './mocks/GnosisReconfigurationPayload.sol';
 
-import { CrosschainTestBase } from './CrosschainTestBase.sol';
+import { CrosschainPayload, CrosschainTestBase } from './CrosschainTestBase.sol';
+
+contract GnosisCrosschainPayload is CrosschainPayload {
+
+    constructor(IPayload _targetPayload, address _bridgeExecutor) CrosschainPayload(_targetPayload, _bridgeExecutor) {}
+
```
Comment (lucas-manuel) [inline test/GnosisCrosschainTest.t.sol:24]:
```suggestion
    constructor(IPayload _targetPayload, address _bridgeExecutor) 
        CrosschainPayload(_targetPayload, _bridgeExecutor) {}
```

File: test/GnosisCrosschainTest.t.sol [hunk 13]
```diff
@@ -4,24 +4,43 @@ pragma solidity ^0.8.0;
...
 
 contract GnosisCrosschainTest is CrosschainTestBase {
 
     IAMB public constant AMB = IAMB(0x75Df5AF045d91108662D8080fD1FEFAd6aA0bb59);
 
     bytes32 public constant MAINNET_CHAIN_ID = bytes32(uint256(1));
 
+    function deployCrosschainPayload(IPayload targetPayload, address bridgeExecutor) public override returns(IPayload) {
+        return IPayload(new GnosisCrosschainPayload(targetPayload, bridgeExecutor));
+    }
```
Comment (lucas-manuel) [inline test/GnosisCrosschainTest.t.sol:45]:
```suggestion
    function deployCrosschainPayload(IPayload targetPayload, address bridgeExecutor) 
        public override returns(IPayload) 
    {
        return IPayload(new GnosisCrosschainPayload(targetPayload, bridgeExecutor));
    }
```

File: test/OptimismCrosschainTest.t.sol [hunk 14]
```diff
@@ -4,16 +4,38 @@ pragma solidity ^0.8.0;
...
+import { IL2BridgeExecutor } from '../src/interfaces/IL2BridgeExecutor.sol';
+
+import { IPayload } from './interfaces/IPayload.sol';
+
+import { CrosschainPayload, CrosschainTestBase } from './CrosschainTestBase.sol';
+
+contract OptimismCrosschainPayload is CrosschainPayload {
+
+    constructor(IPayload _targetPayload, address _bridgeExecutor) CrosschainPayload(_targetPayload, _bridgeExecutor) {}
+
```
Comment (lucas-manuel) [inline test/OptimismCrosschainTest.t.sol:21]:
```suggestion
    constructor(IPayload _targetPayload, address _bridgeExecutor) 
        CrosschainPayload(_targetPayload, _bridgeExecutor) {}
```

File: test/OptimismCrosschainTest.t.sol [hunk 15]
```diff
@@ -4,16 +4,38 @@ pragma solidity ^0.8.0;
...
+}
 
 contract OptimismCrosschainTest is CrosschainTestBase {
 
     address public constant OVM_L2_CROSS_DOMAIN_MESSENGER = 0x4200000000000000000000000000000000000007;
 
+    function deployCrosschainPayload(IPayload targetPayload, address bridgeExecutor) public override returns(IPayload) {
+        return IPayload(new OptimismCrosschainPayload(targetPayload, bridgeExecutor));
+    }
```
Comment (lucas-manuel) [inline test/OptimismCrosschainTest.t.sol:40]:
```suggestion
    function deployCrosschainPayload(IPayload targetPayload, address bridgeExecutor)
        public override returns(IPayload)
    {
        return IPayload(new OptimismCrosschainPayload(targetPayload, bridgeExecutor));
    }
```

File: test/CrosschainTestBase.sol [hunk 16]
```diff
@@ -4,13 +4,13 @@ pragma solidity ^0.8.0;
...
-import { Domain } from 'xchain-helpers/Domain.sol';
 import { BridgedDomain } from 'xchain-helpers/BridgedDomain.sol';
+import { Domain }        from 'xchain-helpers/Domain.sol';
 
 import { IL2BridgeExecutor, IExecutorBase } from '../src/interfaces/IL2BridgeExecutor.sol';
 
-import { IBaseCrosschainForwarder } from './interfaces/IBaseCrosschainForwarder.sol';
-import { IExecutor }                from './interfaces/IExecutor.sol';
+import { IL1Executor } from './interfaces/IL1Executor.sol';
+import { IPayload }  from './interfaces/IPayload.sol';
```
Comment (lucas-manuel) [inline test/CrosschainTestBase.sol]:
```suggestion
import { IL1Executor } from './interfaces/IL1Executor.sol';
import { IPayload }    from './interfaces/IPayload.sol';
```

File: test/CrosschainTestBase.sol [hunk 17]
```diff
@@ -43,23 +81,23 @@ abstract contract CrosschainTestBase is Test  {
     Domain public hostDomain;
     BridgedDomain public bridgedDomain;
 
-    address public forwarder;
     address public bridgeExecutor;
 
+    function deployCrosschainPayload(IPayload targetPayload, address bridgeExecutor) public virtual returns(IPayload);
+
```
Comment (lucas-manuel) [inline test/CrosschainTestBase.sol:85]:
```suggestion
    function deployCrosschainPayload(IPayload targetPayload, address bridgeExecutor) 
        public virtual returns(IPayload);
```

File: test/GnosisCrosschainTest.t.sol [hunk 18]
```diff
@@ -4,24 +4,43 @@ pragma solidity ^0.8.0;
...
 import { IAMB, GnosisBridgeExecutor } from '../src/executors/GnosisBridgeExecutor.sol';
-import { CrosschainForwarderGnosis }  from '../src/forwarders/CrosschainForwarderGnosis.sol';
-import { IL2BridgeExecutor }          from '../src/interfaces/IL2BridgeExecutor.sol';
 
-import { IBaseCrosschainForwarder } from './interfaces/IBaseCrosschainForwarder.sol';
-import { IExecutor }                from './interfaces/IExecutor.sol';
+import { IL2BridgeExecutor } from '../src/interfaces/IL2BridgeExecutor.sol';
+
+import { IL1Executor } from './interfaces/IL1Executor.sol';
+import { IPayload }  from './interfaces/IPayload.sol';
```
Comment (lucas-manuel) [inline test/GnosisCrosschainTest.t.sol]:
```suggestion
import { IL1Executor } from './interfaces/IL1Executor.sol';
import { IPayload }    from './interfaces/IPayload.sol';

```

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
LGTM, mainly just formatting

File: test/ArbitrumCrosschainTest.t.sol (hunk 8 above)
Comment (barrutko) [reply test/ArbitrumCrosschainTest.t.sol:37]: Taken priv.

File: test/CrosschainTestBase.sol (hunk 10 above)
Comment (barrutko) [reply test/CrosschainTestBase.sol:49]: I'd prefer to keep it as is - this is standard format of how we assemble these queue arguments across all repos

**Review by lucas-manuel — APPROVED** [review]

**Review by hexonaut — APPROVED** [review]
LGTM

**Pull request merged**